packages/ra-core/src/controller/list/useSelectAll.tsx

```diff
@@ -0,0 +1,117 @@
+ *   const onSelectAll = useSelectAll({
+ *       selectionModifiers,
+ *       sort: { field: 'title', order: 'ASC' },
+ *       filter: { title: 'foo' },
+ *   });
+ *   const handleClick = () => onSelectAll({
+ *       queryOptions: { meta: { foo: 'bar' } },
+ *       limit: 250,
+ *   });
+ *   return <Button onClick={handleClick}>Select All</Button>;
+ * };
+ */
+export const useSelectAll = ({
+    resource,
+    selectionModifiers,
+    sort,
+    filter,
+}: useSelectAllProps): ((options?: onSelectAllProps) => void) => {
+    const dataProvider = useDataProvider();
+    const queryClient = useQueryClient();
+    const notify = useNotify();
+
+    const onSelectAll = useCallback(
+        async ({ queryOptions = {}, limit = 250 }: onSelectAllProps = {}) => {
+            const { meta, onSuccess, onError } = queryOptions;
+            try {
+                const results = await queryClient.fetchQuery({
+                    queryKey: [
+                        resource,
+                        'getList',
+                        {
+                            resource,
+                            pagination: { page: 1, perPage: limit },
+                            sort,
+                            filter,
+                            meta,
+                        },
+                    ],
+                    queryFn: () =>
+                        dataProvider.getList(resource, {
+                            pagination: {
+                                page: 1,
+                                perPage: limit,
+                            },
+                            sort,
+                            filter,
+                            meta,
+                        }),
+                });
+
+                const allIds = results.data?.map(({ id }) => id) || [];
+                selectionModifiers.select(allIds);
+                if (allIds.length === limit) {
+                    notify('ra.message.too_many_elements', {
+                        messageArgs: { max: limit },
+                        type: 'warning',
+                    });
+                }
+
+                if (onSuccess) {
+                    onSuccess(results);
+                }
+
+                return results.data;
+            } catch (error) {
+                if (onError) {
+                    onError(error);
+                }
+                console.error('Mutation Error: ', error);
+                notify('An error occurred. Please try again.');
+            }
+        },
+        [
+            dataProvider,
+            queryClient,
+            notify,
+            resource,
+            sort,
+            filter,
+            selectionModifiers,
+        ]
+    );
+    return onSelectAll;
+};
+
+export interface useSelectAllProps {
+    resource: string;
+    selectionModifiers: UseRecordSelectionResult[1];
+    sort?: SortPayload;
+    filter?: FilterPayload;
+}
+
+export interface onSelectAllProps<RecordType extends RaRecord = any> {
```

Comment:
I wouldn't use the `Props` suffix as it's not a component. Instead, use `Params`.